Local type aliases

jiripudil opened the pull request:
Resolves phpstan/phpstan#3001. Depends on phpstan/phpdoc-parser#62.

Since Psalm only seems to support class-scoped type aliases, I did the same here. As I've said before, "it seems to be the easiest direction to take and is imo a good first iteration" that neatly follows the 80/20 rule.

I think the implementation still has some rough edges, but it appears to work surprisingly well 😅 looking forward to your thoughts and remarks!

ondrejmirtes commented:
Thank you, this looks very promising. I'm also doing everything with "80/20" in mind 😊

ondrejmirtes commented:
Hi, I'm gonna rebase this myself as I just implemented nested generic bounds on master (https://github.com/phpstan/phpstan-src/commit/e671cc0a5d5c218d4c10aeea9773eb2189c74a34) and I caused quite a few conflicts for you. Stay tuned :)

ondrejmirtes commented on src/PhpDoc/Tag/TypeAliasTag.php:
```diff
@@ -0,0 +1,40 @@
+
+	private string $alias;
+
+	private string $type;
```
Why this isn't a Type but a string?

jiripudil replied:
Because resolving a Type here would be too early, creating a chicken-and-egg problem – to support nested aliases such as `@phpstan-type ListOfAliases AnotherAlias[]`, aliases need to be resolved only after all of them are gathered. Also, imported aliases need to be resolved in the scope of the class that declared them instead of the importing one.

Eventually, I've [wrapped type aliases](https://github.com/phpstan/phpstan-src/pull/460/commits/ee81ee75a09833a73e746e70a7b24f7866e6e6df) into a `TypeAlias` class along with the NameScope in which they are declared, and they are resolved lazily when they are needed.

ondrejmirtes replied:
Great, thank you for the explanation :)

ondrejmirtes commented on src/PhpDoc/Tag/TypeAliasTag.php:
```diff
@@ -0,0 +1,40 @@
+	 * @param mixed[] $properties
+	 * @return TypeAliasTag
+	 */
+	public static function __set_state(array $properties): self
```
This is not needed, objects of classes in `PHPStan\PhpDoc\Tag` are no longer saved in cache.

jiripudil replied:
fixed in [a9c6e17e](https://github.com/phpstan/phpstan-src/pull/460/commits/a9c6e17e9fdbb3d36c98587055c39bf9431b579f)

ondrejmirtes commented on src/Rules/Generics/TemplateTypeCheck.php:
```diff
@@ -79,7 +79,7 @@ public function check(
 				))->build();
 			}
-			if ($this->typeAliasResolver->hasTypeAlias($templateTagName)) {
+			if ($this->typeAliasResolver->hasTypeAlias($templateTagName, $templateTypeScope->getClassName())) {
```
This change should have a test - `@template LocalAlias` should report an error.

jiripudil replied:
Added in [643ffd89](https://github.com/phpstan/phpstan-src/pull/460/commits/643ffd8993cb7bbab2db1fc4c50b2a4978fa6964)

ondrejmirtes commented on src/Type/TypeAliasResolver.php:
```diff
@@ -11,45 +11,124 @@ class TypeAliasResolver
+		}
+
+		if (array_key_exists($aliasNameInClassScope, $this->inProcess)) {
+			throw new \PHPStan\ShouldNotHappenException(sprintf('Circular definition for type alias %s in scope of %s.', $aliasName, $className));
```
Throwing these exceptions causes internal errors. Instead it should be skipped here (`return null`) and reported in a new rule that checks these facts.

jiripudil replied:
Well, I've just preserved the behaviour that was already [there](https://github.com/phpstan/phpstan-src/blob/564b966a31dc339568aaa97c3bee6f195608e7cd/src/PhpDoc/TypeAlias/TypeAliasesTypeNodeResolverExtension.php#L57-L63) :) but ok, I'll try and think of a rule to check for these errors

ondrejmirtes replied:
The difference is that before it was a user error in config but now it's an error in analysed source code (possibly even from third party).

jiripudil replied:
Oh, right, that makes sense :)

ondrejmirtes commented on src/PhpDoc/Tag/TypeAliasImportTag.php:
```diff
@@ -0,0 +1,51 @@
+	 * @param mixed[] $properties
+	 * @return self
+	 */
+	public static function __set_state(array $properties): self
```
Not needed

jiripudil replied:
fixed in [a9c6e17e](https://github.com/phpstan/phpstan-src/pull/460/commits/a9c6e17e9fdbb3d36c98587055c39bf9431b579f)

ondrejmirtes commented on src/Reflection/ClassReflection.php:
```diff
@@ -744,10 +746,38 @@ public function getTypeAliases(): array
+
+			$localAliases = array_map(function (TypeAliasTag $typeAliasTag) use ($importedAliases): string {
+				if (array_key_exists($typeAliasTag->getAlias(), $importedAliases)) {
+					throw new \PHPStan\ShouldNotHappenException(sprintf('Type alias %s overwrites imported type alias of the same name in scope of %s.', $typeAliasTag->getAlias(), $this->getName()));
```
Again, these exceptions should be covered by rules instead.

ondrejmirtes reviewed (changes requested):
I rebased this and added one commit that fixes prefix priority. I really like this! :)

Another test idea: Please try to use an alias right after defining it/importing in class docblock, for example in a `@method` tag.

ondrejmirtes commented:
Oh, wrong rebase, will fix the problem :)

And please, go through the review one-by-one and add each fix as a new commit :)

ondrejmirtes commented:
Alright, giving the branch back to you :) I don't know about the memory limited exceeded in the few failing jobs - it's probably a coincidence that it happened in this branch.

ondrejmirtes commented:
Maybe the memory is leaking somewhere after all...

jiripudil commented:
So, I've fixed some of the issues from your review. I'd say the most notable change is https://github.com/phpstan/phpstan-src/pull/460/commits/210cb621adf8861aae7e95feaa58fbd438ef5664 in which I've wrapped every type alias declaration in a `TypeAlias` class along with its name scope, allowing it to be resolved lazily later. This works around a couple of issues with nested and imported aliases.

I'll try and take a look at the memory issues too, they don't seem to be just a coincidence 🤔

ondrejmirtes commented:
Looking forward to this being finished, it looks awesome already :)

jiripudil commented:
So, I’ve finally found some time to dig into this again. I’ve added the `LocalTypeAliasesRule` that checks for invalid type alias definitions and imports. To make it easier for the rule to detect circular references, they are now resolved to an `ErrorType` instead of null.

---

Regarding the memory issue:

I’ve taken profiles for analysing a single file (essentially the `type-aliases.php` test dataset):

- recent `master` (927d442c8): https://blackfire.io/profiles/4f97b96d-bcbf-495f-ac0a-42e6d0aaa717/graph
- rebased `local-type-aliases`: https://blackfire.io/profiles/6a048c04-1afc-457a-8487-33b68b74d337/graph
- comparison: https://blackfire.io/profiles/compare/83eedd3d-1f08-4321-ab10-a9af108753c6/graph

From what I understand, the local type alias resolution collectivelly consumes ~2.3 MB of memory, most of which (~1.5 MB) is directly accounted for by `ReflectionProvider::hasClass()` lookup, and the remaining ~800 kB are taken by resolving the intersection type in `ExportedTypeAlias` which eventually leads to `ObjectType::getClassReflection()`.

My conclusion is there is no memory leak, it just eats more memory: with the way type alias resolution is implemented, *every* `IdentifierTypeNode` is first attempted to be resolved as a type alias, which adds a significant number of reflection lookups.

That being said, I believe it only has effect on phpstan tests. I assume the reason why only the Windows and PHP >= 7.4 builds are failing is that they don’t use paratest, so all tests are run in the same process and the memory usage accumulates and eventually overflows.

I’ve tried benchmarking the analysis of phpstan itself and even with parallel execution disabled, the results show only a very small increase in memory consumption:

branch | reflection | memory | increase
--- | --- | --- | ---
master@927d442c | runtime | 805.04 MB
master@927d442c | static | 947.04 MB
local-type-aliases | runtime | 815.09 MB | +1,25 %
local-type-aliases | static | 949.09 MB | +0,22 %

How to reproduce:

```
$ php bin/phpstan clear-result-cache
$ php -dmemory_limit=1G bin/phpstan -v
```

```neon
# phpstan.neon

includes: [ build/phpstan.neon ] # for runtime reflection
includes: [ build/phpstan.neon, conf/staticReflection.neon ] # for static reflection

parameters:
	parallel:
		maximumNumberOfProcesses: 1
```

ondrejmirtes commented on conf/config.level0.neon:
```diff
@@ -30,6 +30,7 @@ rules:
 	- PHPStan\Rules\Classes\ExistingClassInTraitUseRule
 	- PHPStan\Rules\Classes\InstantiationRule
 	- PHPStan\Rules\Classes\InvalidPromotedPropertiesRule
+	#- PHPStan\Rules\Classes\LocalTypeAliasesRule
```
Why is it commented out? Although new rules shouldn't be added in minor versions, this is fine as it's for a new feature - no existing code should be affected.

jiripudil replied:
Oh, that's a mistake, thanks for catching it!

ondrejmirtes commented on src/Reflection/ClassReflection.php:
```diff
@@ -769,35 +772,53 @@ public function getTypeAliases(): array
-			$importedAliases = array_map(function (TypeAliasImportTag $typeAliasImportTag): TypeAlias {
+			// prevent circular imports
+			if (array_key_exists($this->getName(), self::$resolvingTypeAliasImports)) {
+				throw new \PHPStan\Type\CircularTypeAliasDefinitionException();
```
Everyone who's calling getTypeAliases() should be prepared to catch this. Wouldn't it be better to just return an empty array? Every situation like that should be caught by the rule anyway.

This is similar to FileTypeMapper that also silently returns empty ResolvedPhpDocBlock.

jiripudil replied:
Under normal circumstances, you should never get this exception from a direct call to getTypeAliases(). It's only thrown if the method is called in recursion a [couple of lines below](https://github.com/phpstan/phpstan-src/pull/460/files/7f066e5fb9ccc7f4106829e775f45ecde86c649d..3519a52cc99046deec8a721a39ccdc0ddd167ebb#diff-1cb8c0f16a3848b23fbd09ac0ce1f5ef06d9e8365d2aded0c6028b8035c78c6fR797-R801), at which point it is caught and transformed into an invalid `TypeAlias`.

Returning `[]` caused some of the circular aliases to be left unresolved, and thus undetectable by the rule without having to reimplement the import logic there.

ondrejmirtes commented on src/PhpDoc/Tag/TypeAliasImportTag.php:
```diff
@@ -0,0 +1,36 @@
+
+	private string $importedAlias;
+
+	private \PHPStan\Type\Type $importedFrom;
```
One idea about this - I think that importedFrom can only ever be IdentifierTypeNode, isn't PHPStan\Type\Type unnecesarily wide type here? It'd be useful if this was just a string with the class name where the tag is from?

jiripudil replied:
I think it would still need to be resolved in the right `NameScope`, the only difference is that it would have to be done later. Currently it happens [early](https://github.com/phpstan/phpstan-src/pull/460/files#diff-649aa72d6962638843a5b8abd489350f92014f519ae86cb9efd388e4ed3dd967R398) and both the [resolver](https://github.com/phpstan/phpstan-src/pull/460/files#diff-1cb8c0f16a3848b23fbd09ac0ce1f5ef06d9e8365d2aded0c6028b8035c78c6fR786) and the [rule](https://github.com/phpstan/phpstan-src/pull/460/files#diff-0d1d8d1217a959a18b46ff0e0b1e1696576e49cd2112fe3b4ac1ecc1d1f56cecR80) check if it's an actual `ObjectType` because it [might not be](https://github.com/phpstan/phpstan-src/pull/460/files#diff-f5beb741d2711011522821608011df3f7ffa1d825db7996eab628378ee24724fR60).

ondrejmirtes replied:
We already eliminated that situation here, right? https://github.com/phpstan/phpdoc-parser/pull/70 so it should be a parse error from the phpdoc-parser

jiripudil replied:
You can still `import from int`, it's a valid identifier so the parser is ok with it, but it's not a valid class name and resolves to `IntegerType`. Which, I agree, is weird. Your suggestion makes sense, we could treat it as a class name and then fail because class `WhateverTheScopeIs\int` does not exist.

ondrejmirtes replied:
I see, `int` is still a valid identifier and the difference is made in phpstan/phpstan in TypeNodeResolver. So parser doesn't know about this.

Yeah, i think it would actually be nicer to have `string` instead of `PHPStan\Type\Type` in TypeAliasImportTag, and call just `NameScope::resolveStringName()` instead of the whole `TypeNodeResolver`. Having to do `instanceof ObjectType` doesn't sit well with me because `TypeAliasImportTag` with IntegerType never makes sense, it shouldn't be created in the first place.

ondrejmirtes commented:
Hi, thank you, these are solid arguments, but it's still a bit suspicious to me.

See the last successful "tests" build on PHP 7.3 on windows: https://github.com/phpstan/phpstan-src/runs/2369417069?check_suite_focus=true (704 MB consumed, 13 minutes 49 seconds)

Here's the same build from your branch: https://github.com/phpstan/phpstan-src/runs/2374358080 (fails after 18 minutes 43 seconds with memory limit exhausted after 81 %)

The memory limit for tests is set to 768 MB in build.xml.

The normal PHPStan analysis is not slower at all, but something weird is happening here. I'm gonna try locally - we're getting really close to merging this :)

ondrejmirtes commented:
My local testing without paratest of the `tests-fast` target:

master: Time: 00:32.972, Memory: 634.50 MB
your branch: Time: 00:30.638, Memory: 829.00 MB

Around 200 MB more is a lot...

ondrejmirtes commented:
Here are logs from both master and this branch of executed tests and consumed memory after each test case class: https://gist.github.com/ondrejmirtes/9465dc4c39440b6afef32c568a6a0669

I'm gonna go through it so I can see if there's a noticable jump in a specific place.

ondrejmirtes commented:
Alright, first big difference: SignatureMapParserTest consumes a lot more in your branch. I'm running it separately like this:

```
cd tests
../vendor/bin/phpunit PHPStan/Reflection/SignatureMap/SignatureMapParserTest.php
```

On master, it consumes 86 MB, on your branch 170 MB.

ondrejmirtes commented:
I pushed two commits into this branch.

What I found out is that in the case of SignatureMapParserTest the memory is consumed after simple `$this->reflectionProvider->hasClass($className)` call. So there might not be a new issue with this branch at all, it's just that there wasn't a reason to ask for that many classes before. And it might be something worth turning off for SignatureMapParser as these classes will never have local aliases anyway :)

ondrejmirtes commented:
Alright, this helps the tests to consume less memory: https://github.com/phpstan/phpstan-src/pull/460/commits/10704635368ae8b6776e14794f53dd18b3c2b8ac And I guess it also helps in runtime a bit...

jiripudil commented:
> So there might not be a new issue with this branch at all, it's just that there wasn't a reason to ask for that many classes before.

Yes, I was thinking something like this could be the culprit in the tests. Thanks for investigating and fixing it!

jiripudil commented on src/Reflection/ClassReflection.php:
```diff
@@ -776,22 +776,19 @@ public function getTypeAliases(): array
 			$typeAliasImportTags = $resolvedPhpDoc->getTypeAliasImportTags();
 			$typeAliasTags = $resolvedPhpDoc->getTypeAliasTags();
 
+			$nameScope = $resolvedPhpDoc->getNullableNameScope() ?? new NameScope(null, []);
```
Is this kind of thing ok? The ResolvedPhpDocBlock internally uses `$nameScope` as non-nullable, but publicly exposes it as nullable. In what circumstances can it actually be null?

ondrejmirtes replied:
I posted what could be done instead as comments :) I guess name scope is null in these situations: https://github.com/phpstan/phpstan-src/blob/927d442c8cd0339f3beb12ff18971fa2b3328ba5/src/Type/FileTypeMapper.php#L105-L111

And also: https://github.com/phpstan/phpstan-src/blob/927d442c8cd0339f3beb12ff18971fa2b3328ba5/src/PhpDoc/ResolvedPhpDocBlock.php#L165

Basically when it knows that NameScope isn't going to be needed (because all tags are already resolved) it doesn't assign it.

jiripudil replied:
Thanks for the explanation! I've fixed that according to your comments, it was obvious once I've read them. I'm gonna go for a run before it's too dark, and come back with a cleaner head to finish this if you or the CI find any other issues :)

ondrejmirtes commented:
While thinking about ObjectType and IdentifierTypeNode, I realized unwanted behaviour:

```php
<?php

/** @phpstan-type int \stdClass */
class Fooooo
{

	/** @param int $a */
	public function doFoo($a)
	{
		\PHPStan\dumpType($a); // should be int, but current is stdClass
	}

}
```

I'm gonna solve it by deleting TypeAliasesTypeNodeResolverExtension and doing the logic in the right place in TypeNodeResolver::resolveIdentifierTypeNode() (after things that are never supposed to be aliased are handled). It should have performance benefits as well.

ondrejmirtes commented on src/Type/TypeAliasResolver.php:
```diff
@@ -98,12 +104,6 @@ private function resolveLocalTypeAlias(string $aliasName, NameScope $nameScope):
-		if (array_key_exists($aliasNameInClassScope, $this->resolvedLocalTypeAliases)) {
-			return $this->resolvedLocalTypeAliases[$aliasNameInClassScope];
-		}
-
```
Is this fine? @jiripudil

jiripudil replied:
Yes, looks fine.

ondrejmirtes commented:
Alright, I think we're green :)

I have two more things to ask from you:

1) Please add another condition checked in the rule - that `@phpstan-type int ShouldNotHappen` isn't allowed. I think you could pass the `int` portion to the `TypeNodeResolver::resolve` and it should result in `ObjectType`.
2) Please send a PR that edits https://github.com/phpstan/phpstan/edit/master/website/src/writing-php-code/phpdoc-types.md and adds documentation about local type aliases :)

Thank you.

jiripudil commented:
Glad to hear that :) I'll add the check and the docs.

Thank you for your help!
